### PR TITLE
Feature: removing players once keep alive expires

### PR DESCRIPTION
**Pull request by wovalle**
Closes https://github.com/lycan-city/werewolf-virtual-cards/issues/33

### CODE REVIEW

File: functions/package.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
     "firebase-functions": "^2.1.0"
   },
   "devDependencies": {
+    "firebase-tools": "^6.1.1",
```
Comment (vetom) [inline functions/package.json:18]: Good one!

File: functions/src/index.ts [hunk 2]
```diff
@@ -1,26 +1,94 @@
...
+        const data = docSnap.data();
+        return {
+          partyId: docSnap.id,
+          uid: Object.keys(data)[0],
+          lastKeepAlive: Object.values(data)[0].toDate(),
+        };
+      })
+    );
+
+  console.log('players with keep alive', playersWithKeepAlive);
```
Comment (vetom) [inline functions/src/index.ts]: Remove all console.log calls so we can keep the log clear for errors.

File: functions/src/index.ts [hunk 3]
```diff
@@ -1,26 +1,94 @@
...
       .delete();
     return change.after.ref.delete();
   });
+
+export const purgePlayers = functions.https.onRequest(async (_, response) => {
+  const db = admin.firestore();
+  db.settings({ timestampsInSnapshots: true }); // So firebase cli stops complaining
+
+  const playersWithKeepAlive = await db
+    .collection('keepAlive')
```
Comment (vetom) [inline functions/src/index.ts:34]: Iterate over all the players online it's an expensive operation, use a `where` to get only the players outside the threshold.

File: functions/src/index.ts [hunk 4]
```diff
@@ -1,26 +1,94 @@
...
+          lastKeepAlive: Object.values(data)[0].toDate(),
+        };
+      })
+    );
+
+  console.log('players with keep alive', playersWithKeepAlive);
+
+  const playersRemoved = [];
+
+  for (let i = 0; i < playersWithKeepAlive.length; i++) {
```
Comment (vetom) [inline functions/src/index.ts:52]: This loop won't be necessary after getting the players filtered.

File: functions/src/index.ts [hunk 5]
```diff
@@ -1,26 +1,94 @@
...
+        .catch(() => null);
+
+      console.log('player party', playerParty);
+
+      const {
+        [player.uid]: playerToRemove,
+        ...remainingPlayers
+      } = playerParty.players;
+
+      const updatedParty = {
```
Comment (vetom) [inline functions/src/index.ts]: Remember to check if `gameInProgress === true` so you kill the player in the game.

**Review by vetom — CHANGES_REQUESTED** [review]
BTW I liked the single commas, can you add the linter rule?

File: functions/src/index.ts (hunk 5 above)
Comment (wovalle) [reply functions/src/index.ts]: Done

File: functions/src/index.ts (hunk 4 above)
Comment (wovalle) [reply functions/src/index.ts:52]: Yes, it will

File: functions/src/index.ts (hunk 3 above)
Comment (wovalle) [reply functions/src/index.ts:34]: With the current data structure is not possible to optimize the query.

File: functions/src/index.ts (hunk 2 above)
Comment (wovalle) [reply functions/src/index.ts]: Updated the logging strategy.

File: functions/src/index.ts (hunk 4 above)
Comment (vetom) [reply functions/src/index.ts:52]: True!

**Review by vetom — APPROVED** [review]

File: functions/src/index.ts [hunk 6]
```diff
@@ -1,26 +1,128 @@
...
+    await db
+      .collection('keepAlive')
+      .doc(player.partyId)
+      .set(remainingKeepAlives);
+  }
+
+  response.send({
+    result: 'OK',
+    playersRemoved,
+    playersActive: playersWithKeepAlive.length - playersRemoved.length,
```
Comment (mamodom) [inline functions/src/index.ts]: I think `activePlayers` sounds better

**Pull request merged**